File: problem119/java/problem119.java
import java.math.BigInteger;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Created by Administrator on 2014/12/24.
 */
public class Problem119 {

    private static int digitsSum(BigInteger x) {
        int sum = 0;
        String string = String.valueOf(x);
        for (int i = 0; i < string.length(); i++) {
            String subString = string.substring(i, i + 1);
            sum += Integer.parseInt(subString);
        }
        return sum;
    }

    public static void main(String[] args) {
        Set<BigInteger> set = new HashSet<>();
        for (int i=2;i<200;i++) {
            for (int j=2;j<200;j++) {
                BigInteger tmp = new BigInteger(String.valueOf(i)).pow(j);
                if (i == digitsSum(tmp)) {
                    set.add(tmp);
                }
            }
        }
        List<BigInteger> list = set.stream().sorted().collect(Collectors.toList());
        System.out.println(list.get(29));
    }
}

File: problem119/Ruby/problem119.rb
def digitsSum(num)
	sum = 0
	num.to_s.each_char do |x|
		sum += x.to_i
	end
	sum
end
#puts digitsSum(123123)

def main()
	list = []
	2.upto(200) do |i|
		2.upto(200) do |j|
			tmp = i ** j
			if i == digitsSum(tmp)
				list.push(tmp)
			end
		end
	end
	list.sort[29]
end

puts main()
